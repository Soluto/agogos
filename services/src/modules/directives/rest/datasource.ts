import {RESTDataSource} from 'apollo-datasource-rest';
import {KeyValue, RestParams} from './types';
import {injectParameters, resolveParameters} from '../../param-injection';
import {RequestContext} from '../../context';
import {RequestInit, Headers, Request} from 'apollo-server-env';
import {GraphQLResolveInfo} from 'graphql';
import {getAuthHeaders} from '../../auth/getAuthHeaders';

type GraphQLArguments = {[key: string]: any};

export class RESTDirectiveDataSource extends RESTDataSource<RequestContext> {
    async doRequest(params: RestParams, parent: any, args: GraphQLArguments, info: GraphQLResolveInfo) {
        const headers = this.parseHeaders(params.headers, parent, args, info);
        const requestInit: RequestInit = {headers, timeout: params.timeoutMs ?? 10000, method: params.method};
        const url = new URL(injectParameters(params.url, parent, args, this.context, info));
        this.addQueryParams(url.searchParams, params.query, parent, args, info);

        const authHeaders = await getAuthHeaders(this.context.authenticationConfig, url.host, this.context.request);
        if (authHeaders != null) {
            headers.append('Authorization', authHeaders.Authorization);
        }

        requestInit.body = args[params.bodyArg || 'input'];
        if (requestInit.body) {
            requestInit.body = JSON.stringify(requestInit.body);
            headers.append('Content-Type', 'application/json');
        }

        const request = new Request(String(url), requestInit);
        const cacheKey = this.cacheKeyFor(request);
        const response = await this.httpCache.fetch(request, {cacheKey});

        return this.didReceiveResponse(response, request);
    }

    parseHeaders(kvs: KeyValue[] | undefined, parent: any, args: GraphQLArguments, info: GraphQLResolveInfo) {
        const headers = new Headers();

        if (typeof kvs !== 'undefined') {
            for (const kv of kvs) {
                headers.append(kv.key, injectParameters(kv.value, parent, args, this.context, info));
            }
        }

        return headers;
    }

    addQueryParams(
        params: URLSearchParams,
        kvs: KeyValue[] | undefined,
        parent: any,
        args: GraphQLArguments,
        info: GraphQLResolveInfo
    ) {
        if (!kvs || kvs.length == 0) return;

        for (const kv of kvs) {
            const parameters = resolveParameters(kv.value, parent, args, this.context, info);
            if (!parameters) {
                params.append(kv.key, kv.value);
                continue;
            }

            for (const originalValue in parameters) {
                const paramValue = parameters[originalValue];
                if (Array.isArray(paramValue)) {
                    for (const elem of paramValue) {
                        params.append(kv.key, elem);
                    }
                } else {
                    params.append(kv.key, paramValue);
                }
            }
        }
    }
}

declare module '../../context' {
    interface ContextDataSources {
        rest: RESTDirectiveDataSource;
    }
}
