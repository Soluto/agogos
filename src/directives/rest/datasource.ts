import {RESTDataSource} from 'apollo-datasource-rest';
import {KeyValue, RestParams} from './types';
import {injectParameters, resolveParameters} from '../../param-injection';
import {RequestContext} from '../../context';
import {RequestInit, Headers} from 'apollo-server-env';

type GraphQLArguments = {[key: string]: any};

export class RESTDirectiveDataSource extends RESTDataSource<RequestContext> {
    doRequest(params: RestParams, parent: any, args: GraphQLArguments) {
        const headers = this.parseHeaders(params.headers, parent, args);
        const requestInit: RequestInit = {headers, timeout: params.timeoutMs};
        const body = params.bodyArg && args[params.bodyArg];
        const url = new URL(injectParameters(params.url, parent, args, this.context));
        this.addQueryParams(url.searchParams, params.query, parent, args);

        const method = params.method?.toUpperCase();

        switch (method) {
            default:
            case 'GET':
                return this.get(url.href, undefined, requestInit);
            case 'DELETE':
                return this.delete(url.href, undefined, requestInit);
            case 'POST':
                return this.post(url.href, body, requestInit);
            case 'PUT':
                return this.put(url.href, body, requestInit);
            case 'PATCH':
                return this.patch(url.href, body, requestInit);
        }
    }

    parseHeaders(kvs: KeyValue[] | undefined, parent: any, args: GraphQLArguments) {
        if (!kvs || kvs.length == 0) return;

        const headers = new Headers();

        for (const kv of kvs) {
            headers.append(kv.key, injectParameters(kv.value, parent, args, this.context));
        }

        return headers;
    }

    addQueryParams(params: URLSearchParams, kvs: KeyValue[] | undefined, parent: any, args: GraphQLArguments) {
        if (!kvs || kvs.length == 0) return;

        for (const kv of kvs) {
            const parameters = resolveParameters(kv.value, parent, args, this.context);
            if (!parameters) {
                params.append(kv.key, kv.value);
                continue;
            }

            for (const originalValue in parameters) {
                const paramValue = parameters[originalValue];
                if (Array.isArray(paramValue)) {
                    for (const elem of paramValue) {
                        params.append(kv.key, elem);
                    }
                } else {
                    params.append(kv.key, paramValue);
                }
            }
        }
    }
}

declare module '../../context' {
    interface ContextDataSources {
        rest: RESTDirectiveDataSource;
    }
}
