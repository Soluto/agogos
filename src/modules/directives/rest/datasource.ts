import {RESTDataSource} from 'apollo-datasource-rest';
import {KeyValue, RestParams} from './types';
import {injectParameters, resolveParameters} from '../../param-injection';
import {RequestContext} from '../../context';
import {RequestInit, Headers, Request} from 'apollo-server-env';

type GraphQLArguments = {[key: string]: any};

export class RESTDirectiveDataSource extends RESTDataSource<RequestContext> {
    async doRequest(params: RestParams, parent: any, args: GraphQLArguments) {
        const headers = this.parseHeaders(params.headers, parent, args);
        const requestInit: RequestInit = {headers, timeout: params.timeoutMs, method: params.method};
        const url = new URL(injectParameters(params.url, parent, args, this.context));
        this.addQueryParams(url.searchParams, params.query, parent, args);

        requestInit.body = args[params.bodyArg || 'input'];
        if (requestInit.body) {
            requestInit.body = JSON.stringify(requestInit.body);
            headers.append('Content-Type', 'application/json');
        }

        const request = new Request(String(url), requestInit);
        const cacheKey = this.cacheKeyFor(request);
        const response = await this.httpCache.fetch(request, {cacheKey});

        return this.didReceiveResponse(response, request);
    }

    parseHeaders(kvs: KeyValue[] | undefined, parent: any, args: GraphQLArguments) {
        const headers = new Headers();

        if (typeof kvs !== 'undefined') {
            for (const kv of kvs) {
                headers.append(kv.key, injectParameters(kv.value, parent, args, this.context));
            }
        }

        return headers;
    }

    addQueryParams(params: URLSearchParams, kvs: KeyValue[] | undefined, parent: any, args: GraphQLArguments) {
        if (!kvs || kvs.length == 0) return;

        for (const kv of kvs) {
            const parameters = resolveParameters(kv.value, parent, args, this.context);
            if (!parameters) {
                params.append(kv.key, kv.value);
                continue;
            }

            for (const originalValue in parameters) {
                const paramValue = parameters[originalValue];
                if (Array.isArray(paramValue)) {
                    for (const elem of paramValue) {
                        params.append(kv.key, elem);
                    }
                } else {
                    params.append(kv.key, paramValue);
                }
            }
        }
    }
}

declare module '../../context' {
    interface ContextDataSources {
        rest: RESTDirectiveDataSource;
    }
}
