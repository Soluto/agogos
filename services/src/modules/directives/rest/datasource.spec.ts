import 'jest-fetch-mock';
import {InMemoryLRUCache} from 'apollo-server-caching';
import {RESTDirectiveDataSource} from './datasource';
jest.mock('../../logger');

const emptyContext = {
    authenticationConfig: {
        getUpstreamByHost() {},
    },
    request: {headers: {}},
};
describe('REST directive data source', () => {
    let ds: RESTDirectiveDataSource;

    beforeEach(() => {
        fetchMock.resetMocks();
        ds = new RESTDirectiveDataSource(fetchMock as any);
        ds.initialize({context: emptyContext as any, cache: new InMemoryLRUCache()});
    });

    it('Defaults to GET', async () => {
        await ds.doRequest({url: 'http://somewhere'}, null, {}, {} as any);

        expect(fetchMock).toHaveBeenCalledTimes(1);
        const req = fetchMock.mock.calls[0][0] as Request;
        expect(req.url).toBe('http://somewhere/');
        expect(req.method).toBe('GET');
    });

    it('Includes query parameters, static ones & from params & arrays from params', async () => {
        await ds.doRequest(
            {
                url: 'http://somewhere',
                query: [
                    {key: 'mykey', value: 'myvalue'},
                    {key: 'owner', value: '{args.name}'},
                    {key: 'child', value: '{source.children}'},
                ],
            },
            {children: ['one', 'two', 3]},
            {name: 'aviv'},
            {} as any
        );

        expect(fetchMock).toHaveBeenCalledTimes(1);
        const req = fetchMock.mock.calls[0][0] as Request;
        expect(req.url).toBe('http://somewhere/?mykey=myvalue&owner=aviv&child=one&child=two&child=3');
    });

    it('Includes headers', async () => {
        await ds.doRequest(
            {
                url: 'http://somewhere',
                headers: [
                    {key: 'mykey', value: 'myvalue'},
                    {key: 'owner', value: '{args.name}'},
                ],
            },
            null,
            {name: 'aviv'},
            {} as any
        );

        expect(fetchMock).toHaveBeenCalledTimes(1);
        const req = fetchMock.mock.calls[0][0] as Request;
        expect(req.headers).toStrictEqual(new Headers({mykey: 'myvalue', owner: 'aviv'}));
        expect(req.url).toBe('http://somewhere/');
        expect(req.method).toBe('GET');
    });

    it('Skips headers and query params who resolve to undefined', async () => {
        await ds.doRequest(
            {
                url: 'http://somewhere',
                query: [{key: 'name', value: '{source.firstName}'}],
                headers: [{key: 'name', value: '{source.firstName}'}],
            },
            {firstName: undefined},
            {},
            {} as any
        );

        expect(fetchMock).toHaveBeenCalledTimes(1);
        const req = fetchMock.mock.calls[0][0] as Request;
        expect(req.url).toBe('http://somewhere/');
    });

    describe('Supported http methods dispatch with the correct http method', () => {
        const methods = ['GET', 'DELETE', 'POST', 'PUT', 'PATCH'];

        for (const method of methods) {
            it(method, async () => {
                await ds.doRequest(
                    {
                        url: 'http://somewhere',
                        method,
                    },
                    null,
                    {},
                    {} as any
                );

                expect(fetchMock).toHaveBeenCalledTimes(1);
                const req = fetchMock.mock.calls[0][0] as Request;
                expect(req.method).toBe(method);
            });
        }
    });

    it('Adds stringified JSON body', async () => {
        await ds.doRequest(
            {
                url: 'http://somewhere',
                method: 'POST',
                bodyArg: 'input',
            },
            null,
            {input: {name: 'aviv'}},
            {} as any
        );

        expect(fetchMock).toHaveBeenCalledTimes(1);
        const req = fetchMock.mock.calls[0][0] as Request;
        expect(req.url).toBe('http://somewhere/');
        expect(req.method).toBe('POST');
        expect(req.headers.get('Content-Type')).toBe('application/json');
        expect(await req.text()).toBe('{"name":"aviv"}');
    });
});
