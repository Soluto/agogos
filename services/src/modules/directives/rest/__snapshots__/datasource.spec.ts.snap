// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`REST directive data source Does not send request if required headers are empty 1`] = `[GraphQLError: mykey header is required]`;

exports[`REST directive data source Does not send request if required query parameters are missing values 1`] = `[GraphQLError: field1 query parameter is required]`;
